src/lib.rs

```diff
@@ -338,7 +343,20 @@ impl<B: MysqlShim<W>, R: Read, W: Write> MysqlIntermediary<B, R, W> {
                     }];
                     writers::write_column_definitions(cols, &mut self.writer, true)?;
                 }
-                Command::Init(_) | Command::Ping => {
+                Command::Init(schema) => {
+                    let result = self.shim.on_init(
+                        ::std::str::from_utf8(schema)
+                                    .map_err(|e| io::Error::new(io::ErrorKind::InvalidData, e))?
+                                    );
+                    match result {
+                        Ok(_x) => writers::write_ok_packet(&mut self.writer, 0, 0, StatusFlags::empty())?,
+                        Err(_e) => {
```

Comment:
Are we expecting that the user won't ever want to return some kind of useful error from `on_init` that should be communicated back?

Reply:
I've made another commit that allows the caller to shove a message into the io error however it looks a little janky. Maybe I should figure out how to allow the caller to pass a mysql::errro::MySqlError back. What do you think?

Reply:
Yeah, I think that'd be better. Not entirely sure how to do that though. We _could_ do a `Result<Result<>>` type setup, but that's pretty janky too. The "real" way to do this would be to provide `on_init` with a "writer" like the other `on_` methods that the implementor can use either to write an ok or an error response. In this particular case, that writer can be pretty straightforward since all the user can do is write an ok or an error :p

Reply:
I think it's looking better with the InitWriter<W>. I'm having a heck of a time figuring out how to write good tests for it though - the mysql_simple crate doesn't implement a select_db() [1] so it's kind of impossible to trigger this code with it :) I might need to connect my own TcpStream to the port to shove the packet in but I'd rather not. I'll think about it some more tomorrow. 

[1] https://github.com/blackbeam/rust-mysql-simple/issues/170